File: daily/validParentheses.cpp
class Solution {
public:
    bool isValid(string s) {
        if(s.size() < 2) return false;
        map<char, char> bracs = {
            {')','('},
            {'}','{'},
            {']','['}
        };
        map<char, bool> entry = {
            {'(', true},
            {'{', true},
            {'[', true},
            {')', false},
            {']', false},
            {'}', false},
        };
        
        stack<char> c;
        
        for(int i=0;i<s.size();i++){
            if(entry[s[i]]){
                c.push(s[i]);
                continue;
            }
            if(c.empty()) return false;
            if(c.top() != bracs[s[i]]) return false;
            c.pop();
        }
        return c.empty();
    }
};

File: daily/100daycode/validParentheses.py
class Solution(object):
    def isValid(self, s):
        """
        :type s: str
        :rtype: bool
        """
        stack = []
        seq = {
            ')': '(',
            ']': '[',
            '}': '{'
        }
        for p in s:
            if p in seq:
                if stack and stack[-1] == seq[p]:
                    stack.pop()
                else:
                    return False
            else:
                stack.append(p)
        return True if not stack else False
        
